pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -113,6 +126,108 @@ func (p *provisioningAdmitter) Admit(request *admission.Request) (*admissionv1.A
 	return response, nil
 }
 
+// validateAgentEnvVars will validate the provisioning cluster object's AgentEnvVars, ensuring that the
+// "CATTLE_AGENT_VAR_DIR" env var is not allowed to be updated.
+func (p *provisioningAdmitter) validateAgentEnvVars(request *admission.Request, oldCluster, newCluster *v1.Cluster) *metav1.Status {
+	if newCluster.Spec.RKEConfig == nil {
+		return nil
+	}
+	if request.Operation == admissionv1.Create {
+		if slices.ContainsFunc(newCluster.Spec.AgentEnvVars, func(envVar rkev1.EnvVar) bool {
+			return envVar.Name == "CATTLE_AGENT_VAR_DIR"
+		}) {
+			return &metav1.Status{
+				Status:  "Failure",
+				Message: `"CATTLE_AGENT_VAR_DIR" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`,
+				Reason:  metav1.StatusReasonBadRequest,
+				Code:    http.StatusBadRequest,
+			}
+		}
+		return nil
+	}
+	if request.Operation == admissionv1.Update {
+		var oldCount, newCount int
+		for _, envVar := range oldCluster.Spec.AgentEnvVars {
+			if envVar.Name == "CATTLE_AGENT_VAR_DIR" {
+				oldCount++
+			}
+		}
+		for _, envVar := range newCluster.Spec.AgentEnvVars {
+			if envVar.Name == "CATTLE_AGENT_VAR_DIR" {
+				newCount++
+			}
+		}
+		if newCount > oldCount {
```

Comment:
I'm not sure that this works as intended. From what I can see here, I could change this value as long as I added/removed in a single step. For example, if I change `CATTLE_AGENT_VAR_DIR` from `/var/lib/rke2` to `/var/lib/new`, that would work.

That being said, I'm not 100% sure that this is intentional - open to hearing otherwise.

Reply:
Honestly it's not super clear but it is functionally correct: this function is making sure that new `CATTLE_AGENT_VAR_DIR` vars are only allowed to be added, not removed. The other function is supposed to do validation, which is covered by this case:
```
	oldEnvVars := slices.DeleteFunc(slices.Clone(oldCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
		return envVar.Name != systemAgentVarDirEnvVar
	})
	// will be true up until rancher performs one time migration to set SystemAgent data directory
	if len(oldEnvVars) > 0 {
		newEnvVars := slices.DeleteFunc(slices.Clone(newCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
			return envVar.Name != systemAgentVarDirEnvVar
		})
		if reflect.DeepEqual(newEnvVars, oldEnvVars) && oldCluster.Spec.RKEConfig.DataDirectories.SystemAgent == newCluster.Spec.RKEConfig.DataDirectories.SystemAgent {
			// rancher migration has not been performed yet
			return admission.ResponseAllowed()
		}
		if newCluster.Spec.RKEConfig.DataDirectories.SystemAgent != oldEnvVars[len(oldEnvVars)-1].Value {
			return admission.ResponseBadRequest(
				fmt.Sprintf(`"%s" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`, systemAgentVarDirEnvVar))
		}
```
so if we were defining the env var previously, and we haven't changed either the env var or data directory field in this update, assume it's fine (actually as I look at it we shouldn't allow this without validating the other data directories but you get the idea).

Reply:
I might be missing something, but the current method will only stop me from adding/removing the env var right? So if I just change it, it's allowed. From your reply we don't want that, it needs to be the same value as before correct?